packages/babel-plugin-transform-typescript/src/enum.ts

```diff
@@ -131,109 +137,168 @@ const enumSelfReferenceVisitor = {
+          true,
+        );
+        value = t.binaryExpression("+", t.numericLiteral(1), lastRef);
         seen.set(name, undefined);
       }
-    } else if (typeof constValue === "number") {
-      constValue += 1;
-      value = t.numericLiteral(constValue);
-      seen.set(name, constValue);
-    } else if (typeof constValue === "string") {
-      throw path.buildCodeFrameError("Enum member must have initializer.");
-    } else {
-      // create dynamic initializer: 1 + ENUM["PREVIOUS"]
-      const lastRef = t.memberExpression(
-        t.cloneNode(path.node.id),
-        t.stringLiteral(lastName),
-        true,
-      );
-      value = t.binaryExpression("+", t.numericLiteral(1), lastRef);
-      seen.set(name, undefined);
-    }
 
-    lastName = name;
-    return [name, value];
-  });
+      lastName = name;
+      return [name, value];
+    }) as Array<[name: string, value: t.Expression]>,
+  };
 }
 
 // Based on the TypeScript repository's `evalConstant` in `checker.ts`.
 function evaluate(
-  expr: t.Node,
+  path: NodePath,
   seen: PreviousEnumMembers,
-): number | string | typeof undefined {
-  return evalConstant(expr);
+): number | string | undefined {
+  return evalConstant(path);
 
-  function evalConstant(expr: t.Node): number | typeof undefined {
+  function evalConstant(path: NodePath): number | string | undefined {
+    const expr = path.node;
     switch (expr.type) {
+      case "MemberExpression": {
+        const obj = expr.object;
+        const prop = expr.property;
+        if (
+          !types.isIdentifier(obj) ||
+          (expr.computed
+            ? !types.isStringLiteral(prop)
+            : !types.isIdentifier(prop))
+        ) {
+          return;
+        }
+        const bindingIdentifier = path.scope.getBindingIdentifier(obj.name);
+        const data = ENUMS.get(bindingIdentifier);
+        if (!data) return;
+        // @ts-expect-error checked above
+        return data.get(prop.computed ? prop.value : prop.name);
+      }
+
       case "StringLiteral":
         return expr.value;
       case "UnaryExpression":
-        return evalUnaryExpression(expr);
+        return evalUnaryExpression(path as NodePath<t.UnaryExpression>);
       case "BinaryExpression":
-        return evalBinaryExpression(expr);
+        return evalBinaryExpression(path as NodePath<t.BinaryExpression>);
       case "NumericLiteral":
         return expr.value;
       case "ParenthesizedExpression":
-        return evalConstant(expr.expression);
-      case "Identifier":
-        return seen.get(expr.name);
-      case "TemplateLiteral":
+        return evalConstant(path.get("expression"));
+      case "Identifier": {
+        let value = seen.get(expr.name);
+        if (value === undefined) {
+          value = evaluateIdentifier(path);
+          if (value !== undefined) {
+            seen.set(expr.name, value);
+          }
+        }
+        return value;
+      }
+      case "TemplateLiteral": {
```

Comment:
What is the difference between the code in this `case` and just using `path.evaluate()`? It seems to me that this code doesn't access the `ENUMS` weakmap, so they should be equivalent.

Reply:
The main inconsistency comes from `evaluateIdentifier`, and overall, `path.evaluate()` will be more aggressive, while ts will be safer. However ts I guess the officially supported `const var` should always work fine.